### PR TITLE
Implement server-side single-repo context fetching for RFC 969

### PR DESCRIPTION
**Pull request by rafax**
Implements server-side part of AI-132, exposing a single-repo context endpoint backed by Zoekt.

Differences from the current Enterprise endpoint (that is left untouched):

- supports a single repo only
- supports passing a repo name, not Sourcegraph-assigned repo ID
- has a shorter, 5s timeout
- reports partial errors (if we hit issues converting search results to git results, or if some retrievers time out)

## Test plan

- tested locally with 
```
{
  chatContext(
    query: "squirrel"
    repo: "github.com/sourcegraph/sourcegraph"
    interactionId: "foo"
  ) {
    contextItems {
      score
      retriever
      item {
        ... on FileChunkContext {
          blob {
            path
            repository {
              id
              name
            }
            commit {
              oid
            }
            url
          }
          startLine
          endLine
          chunkContent
        }
      }
    }
    partialErrors
  }
}

```

### CODE REVIEW

File: cmd/frontend/internal/context/resolvers/context.go [hunk 1]
```diff
@@ -53,6 +58,54 @@ type Resolver struct {
 	cohereConfig        *schema.CodyRerankerCohere
 }
 
+func (r *Resolver) ChatContext(ctx context.Context, args graphqlbackend.ChatContextArgs) (graphqlbackend.ChatContextResolver, error) {
+	if !dotcom.SourcegraphDotComMode() {
+		return nil, errors.New("chat context is not available on Sourcegraph.com")
```
Comment (janhartman) [inline cmd/frontend/internal/context/resolvers/context.go:63]: I didn't look into the implementation but the naming of the function implies that this is only available on dotcom. I think we should either remove the negation in the if clause or change the error message to reflect the behavior.

File: cmd/frontend/graphqlbackend/cody_context.graphql [hunk 2]
```diff
@@ -90,6 +90,28 @@ extend type Query {
...
+        """
+        query: String!
+        """
+        The repository to search.
+        """
+        repo: String!
+        """
+        Interaction ID.
+        """
+        interactionId: String!
```
Comment (janhartman) [inline cmd/frontend/graphqlbackend/cody_context.graphql:109]: what does this mean? just a general ID of the call we're making?

File: cmd/frontend/graphqlbackend/cody_context.graphql [hunk 3]
```diff
@@ -191,3 +213,39 @@ type RankingResponse {
...
+    """
+    Context items fetched for this chat interaction.
+    """
+    contextItems: [RetrieverContextItem!]!
+    """
+    Partial errors that happened while fetching context items (but did not block the whole request).
+    """
+    partialErrors: [String!]!
+    """
+    Partial errors that happened while fetching context items (but did not block the whole request).
```
Comment (janhartman) [inline cmd/frontend/graphqlbackend/cody_context.graphql:230]: This description is wrong (just a copy of the above one).

File: cmd/frontend/internal/context/resolvers/context.go [hunk 4]
```diff
@@ -53,6 +58,54 @@ type Resolver struct {
...
+	}
+	// Set a more aggressive timeout for this request so slow experimental retrievers won't block client
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	repo, err := r.db.Repos().GetByName(ctx, api.RepoName(args.Repo))
+	if err != nil {
+		return nil, err
+	}
+	fileChunks, err := r.contextClient.GetCodyContext(ctx, codycontext.GetContextArgs{
```
Comment (janhartman) [inline cmd/frontend/internal/context/resolvers/context.go:72]: Can you maybe add some boilerplate to generalize this to more retrievers? Just have a list of retrievers and call them in parallel, even though it's just zoekt for now.

**Review by janhartman — APPROVED** [review]
Nice!

File: cmd/frontend/internal/context/resolvers/context.go (hunk 1 above)
Comment (rafax) [reply cmd/frontend/internal/context/resolvers/context.go:63]: Good catch, I'll use the same function as for other endpiints

File: cmd/frontend/graphqlbackend/cody_context.graphql (hunk 2 above)
Comment (rafax) [reply cmd/frontend/graphqlbackend/cody_context.graphql:109]: It's a single id that should be shared by intentDetection/context fetching/context ranking/context recording - we expect it to be the same for multiple backend requests, so I can't call it requestId.

Comment (rafax) [reply cmd/frontend/graphqlbackend/cody_context.graphql:109]: added more context

File: cmd/frontend/graphqlbackend/cody_context.graphql (hunk 3 above)
Comment (rafax) [reply cmd/frontend/graphqlbackend/cody_context.graphql:230]: Sure

File: cmd/frontend/internal/context/resolvers/context.go (hunk 4 above)
Comment (rafax) [reply cmd/frontend/internal/context/resolvers/context.go]: PTAL if this looks ok now

File: cmd/frontend/internal/context/resolvers/context.go [hunk 5]
```diff
@@ -53,6 +68,50 @@ type Resolver struct {
...
+	success := false
+	iter.ForEach(retrievers, func(f *retrieverFunc) {
+		items, pe, err := (*f)(ctx, repo, args.Query, r)
+		if err != nil {
+			completeErrors = append(completeErrors, err)
+			return
+		}
+		success = true
+		res.contextItems = append(res.contextItems, items...)
+		partialErrors = append(partialErrors, pe...)
```
Comment (janhartman) [inline cmd/frontend/internal/context/resolvers/context.go:95]: This isn't thread safe - should we map (concurrently) first and then aggregate?

Comment (rafax) [reply cmd/frontend/internal/context/resolvers/context.go:95]: Fair, let me use `iter.Map`

**Review by janhartman — APPROVED** [review]
Nice, thanks for the fix!

**Pull request merged**